profiler/profile.go

```diff
@@ -112,43 +189,59 @@ func (b *batch) addProfile(p *profile) {
 	b.profiles = append(b.profiles, p)
 }
 
-func (p *profiler) runProfile(t ProfileType) (*profile, error) {
-	switch t {
-	case HeapProfile:
-		return heapProfile(p.cfg)
-	case CPUProfile:
-		return cpuProfile(p.cfg)
-	case MutexProfile:
-		return mutexProfile(p.cfg)
-	case BlockProfile:
-		return blockProfile(p.cfg)
-	case GoroutineProfile:
-		return goroutineProfile(p.cfg)
-	case expGoroutineWaitProfile:
-		return goroutineWaitProfile(p.cfg)
-	case MetricsProfile:
-		return p.collectMetrics()
-	default:
-		return nil, errors.New("profile type not implemented")
+func (p *profiler) runProfile(t ProfileType) ([]*profile, error) {
+	var (
+		pt    = t.collector()
+		start = now()
+	)
+	data, err := pt.Collect(pt, p)
+	if err != nil {
+		return nil, err
 	}
-}
 
-// writeHeapProfile writes the heap profile; replaced in tests
-var writeHeapProfile = pprof.WriteHeapProfile
+	profs := []*profile{{
+		name: pt.Filename,
+		data: data,
+	}}
 
-func heapProfile(cfg *config) (*profile, error) {
-	var buf bytes.Buffer
-	start := now()
-	if err := writeHeapProfile(&buf); err != nil {
-		return nil, err
+	if pt.Delta {
+		p1, err := pprofile.Parse(bytes.NewReader(data))
+		if err != nil {
+			return nil, fmt.Errorf("delta prof parse: %v", err)
+		}
+
+		if p0 := p.prev[t]; p0 != nil {
+			deltaProf, err := pprofile.Merge([]*pprofile.Profile{p0, p1})
+			if err != nil {
+				return nil, fmt.Errorf("delta prof merge: %v", err)
+			}
+			deltaBuf := &bytes.Buffer{}
+			if err := deltaProf.Write(deltaBuf); err != nil {
+				return nil, fmt.Errorf("delta prof write: %v", err)
+			}
+
+			// TODO(fg) do we need to modify TimeNanos here?
+			// https://github.com/golang/go/commit/2ff1e3ebf5de77325c0e96a6c2a229656fc7be50#diff-94594f8f13448da956b02997e50ca5a156b65085993e23bbfdda222da6508258R303-R304
+
+			profs = append(profs, &profile{
+				// TODO(fg) are those good filenames? Is there a better way to flag
+				// these profiles for the backend?
+				name: "delta-" + pt.Filename,
+				data: deltaBuf.Bytes(),
+			})
+		}
+
```

Comment:
I think the same goes here. There is plenty of spacing above not to clutter the code so we can get rid of this empty line.

Reply:
Are the two places you pointed out the main ones you think should change? From re-reading this code I felt 9e1120ccdd02b31f6bb95788bbdb81d48cff7f8e was the most offensive one and removed it.